presto-geospatial/src/main/java/com/facebook/presto/plugin/geospatial/GeoFunctions.java

```diff
@@ -378,6 +388,100 @@ public static Slice stCentroid(@SqlType(GEOMETRY_TYPE_NAME) Slice input)
         return serialize(geometry.getCentroid());
     }
 
+    @Description("Returns the Point value that is the mathematical centroid of a Spherical Geography")
+    @ScalarFunction("ST_Centroid")
+    @SqlType(SPHERICAL_GEOGRAPHY_TYPE_NAME)
+    public static Slice stSphericalCentroid(@SqlType(SPHERICAL_GEOGRAPHY_TYPE_NAME) Slice input)
+    {
+        OGCGeometry geometry = EsriGeometrySerde.deserialize(input);
+        if (geometry.isEmpty()) {
+            return null;
+        }
+        // TODO: add support for other types e.g. POLYGON
+        validateSphericalType("ST_Centroid", geometry, EnumSet.of(POINT, MULTI_POINT));
+        if (! (geometry instanceof OGCGeometryCollection)) {
+            return input;
+        }
+
+        OGCGeometryCollection geometryCollection = (OGCGeometryCollection)geometry;
+        for (int i = 0; i < geometryCollection.numGeometries(); i++) {
+            OGCGeometry g = geometryCollection.geometryN(i);
+            validateSphericalType("ST_Centroid", g, EnumSet.of(POINT));
+            Point p = (Point) g.getEsriGeometry();
+            checkLatitude(p.getX());
+            checkLongitude(p.getY());
+        }
+
+        Point centroid = null;
+        if (geometryCollection.numGeometries() == 1) {
+            return input;
+        }else if(geometryCollection.numGeometries() == 2) {
+            // 2 points
+            CartesianPoint cp1 = new CartesianPoint((Point)geometryCollection.geometryN(0).getEsriGeometry());
+            CartesianPoint cp2 = new CartesianPoint((Point)geometryCollection.geometryN(1).getEsriGeometry());
+            centroid = CartesianPoint.convertToLatLong(
+                    (cp1.x + cp2.x)/2,
+                    (cp1.y + cp2.y)/2,
+                    (cp1.z + cp2.z)/2
+            );
+        }else if(geometryCollection.numGeometries() == 3) {
+            // Triangle
+            CartesianPoint cp1 = new CartesianPoint((Point)geometryCollection.geometryN(0).getEsriGeometry());
+            CartesianPoint cp2 = new CartesianPoint((Point)geometryCollection.geometryN(1).getEsriGeometry());
+            CartesianPoint cp3 = new CartesianPoint((Point)geometryCollection.geometryN(2).getEsriGeometry());
+            double x3D = (cp1.x + cp2.x + cp3.x)/3;
+            double y3D = (cp1.y + cp2.y + cp3.y)/3;
+            double z3D = (cp1.z + cp2.z + cp3.z)/3;
+            double centroidVectorLength = Math.sqrt(x3D*x3D + y3D*y3D + z3D*z3D);
+            centroid = CartesianPoint.convertToLatLong(
+                    x3D / centroidVectorLength,
+                    y3D / centroidVectorLength,
+                    z3D / centroidVectorLength
+            );
+        }else {
```

Comment:
Yes, it would definitely be simpler, but there are degenerate cases that would give the wrong answer e.g. opposite point which would resolve to the center of the earth.

Reply:
Updated to take the average of the points